Electron Sample Using System browser and Custom URL Scheme (#4972)
AzureAD/microsoft-authentication-library-for-js: samples/msal-node-samples/ElectronSystemBrowserApp/README.md

# MSAL Node Electron Sample Using System browser and Custom URL Scheme

An Electron application built with TypeScript that uses the MSAL Node library to acquire and store access tokens to authenticate with the Microsoft Graph API.

## Should I use this sample

> Understanding the way MSAL Node is used in this sample will help you if you're interested in building an Electron.js application to authenticate and acquire tokens using the system browser and custom URL scheme.

Reviewer: Can add a few more lines here? What advantage does using the system browser provide?

```

When the dependencies have been installed, you can run the sample application by using the following command, after which the Electron application should start.

```bash
cd samples/msal-node-samples/standalone-samples/ElectronSystemBrowserApp/redirect
$ npm start
```

```bash
cd samples/msal-node-samples/standalone-samples/ElectronSystemBrowserApp
npm start
```

## About the code

### Custom URL Scheme

To demonstrate best security practices, this Electron sample uses a custom URL scheme `msal4b0db8c2-9f26-4417-8bde-3f0e3656f8e0://auth` that will launch the app when the URL with that scheme is visited.

Using this method, the desktop application starts the [Authorization code flow with PKCE flow](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow) as normal by lunching the system browser, authenticating the user, and redirecting with the authorization code in the browser.

On receiving the authorization code from the **AzureAD**, the browser will launch the application with the custom URL scheme and the code as query parameters. The desktop will parse the authorization code to authenticate the user with **MSAL**.

#### Redirect page

> We used a node server to host a static web page `.\redirect\index.html` to serve as a redirect page. The page will receive the authorization code as query parameter from **AzureAD** and on button click will trigger the Custom URL Scheme `msal4b0db8c2-9f26-4417-8bde-3f0e3656f8e0://auth` to forward the authorization code to the electron application.

Reviewer: Perhaps we can mention hosting this page on Azure Storage static websites or something similar?

```html
 <script>
         window.addEventListener('DOMContentLoaded', function () {

>             var redirectUri = "msal4b0db8c2-9f26-4417-8bde-3f0e3656f8e0://auth";

Reviewer: any way to read redirectUri as a parameter from somewhere, or do we mention that they need to update this in the readme?

            if (window.location.search) {

                redirectUri += window.location.search;
                window.history.pushState({}, document.title, "/");
            }

            document.getElementById('continueButton').onclick = function () {
                window.location.href = redirectUri
            };
         });
    </script>
```

### Deep linking

> To register the custom URL scheme and to link to the electron app. we Will have to utilize  `setAsDefaultProtocolClient()` API provided by electron.

Reviewer: ```suggestion
To register the custom URL scheme and to link to the electron app instance, we will utilize the `setAsDefaultProtocolClient()` API provided by electron.
```

For more information about using deep linking with electron, check [electron documentation](https://www.electronjs.org/docs/latest/tutorial/launch-app-from-url-in-another-ap)

```typescript
private static setDefaultProtocol(): void {
        if (process.defaultApp) {
            if (process.argv.length >= 2) {
                Main.application.setAsDefaultProtocolClient(
                    authConfig.customProtocol.name, // e.g msal4b0db8c2-9f26-4417-8bde-3f0e3656f8e
                    process.execPath,
                    [path.resolve(process.argv[1])]
                );
            } else {
                Main.application.setAsDefaultProtocolClient(
                    authConfig.customProtocol.name // e.g msal4b0db8c2-9f26-4417-8bde-3f0e3656f8e
                );
            }
        }
    }
```

> Next, we tell the application how to handle an event where an external protocol is clicked. This code will differ from Windows compared to macOS and Linux.

Reviewer: ```suggestion
Next, we tell the application how to handle events that invoke the application instance (such as navigation to the custom url scheme in browser). This code will differ between OSs.
```

5. In the Redirect URI section,  enter the following redirect URI `http://localhost:8080/`.
6. Copy the exact `redirectUri` value (`msal4b0db8c2-9f26-4417-8bde-3f0e3656f8e0://auth` is the default if you don't want to change the configuration) into the input box.
7. Click "Configure"

### Executing the application

Once you are in the sample application directory and you've configured the sample application to match your App Registration and registered Redirect URI, you can install all of the dependencies by running:

```bash
npm install
```

When the dependencies have been installed, you can run the sample application by using the following command, after which the Electron application should start.

```bash
cd samples/msal-node-samples/standalone-samples/ElectronSystemBrowserApp/redirect
$ npm start
```

```bash
cd samples/msal-node-samples/standalone-samples/ElectronSystemBrowserApp
npm start
```

## About the code

### Custom URL Scheme

To demonstrate best security practices, this Electron sample uses a custom URL scheme `msal4b0db8c2-9f26-4417-8bde-3f0e3656f8e0://auth` that will launch the app when the URL with that scheme is visited.

> Using this method, the desktop application starts the [Authorization code flow with PKCE flow](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow) as normal by lunching the system browser, authenticating the user, and redirecting with the authorization code in the browser.

Reviewer 2: ```suggestion
Using this method, the desktop application starts the [Authorization code flow with PKCE flow](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow) as normal by launching the system browser, authenticating the user, and redirecting with the authorization code in the browser.
```

On receiving the authorization code from the **AzureAD**, the browser will launch the application with the custom URL scheme and the code as query parameters. The desktop will parse the authorization code to authenticate the user with **MSAL**.

#### Redirect page

> We used a node server to host a static web page [index.html](./redirect/public/index.html) to serve as a redirect page. The page will receive the authorization code as query parameter from **AzureAD** and on button click will trigger the Custom URL Scheme `msal4b0db8c2-9f26-4417-8bde-3f0e3656f8e0://auth` in the [authConfig.js](./redirect/public/config/authConfig.json) to forward the authorization code to the electron application. Deploying the static web page and using it as a redirect page for the electron app is recommended. For more information on hosting a static website on **Azure Storage**, please see [Host a static website on Blob Storage](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-blob-static-website-host)

Reviewer: ```suggestion
We used a node server to host a static web page [index.html](./redirect/public/index.html) to serve as a redirect page. The page will receive the authorization code as query parameter from **AzureAD** and on button click will trigger the Custom URL Scheme `msal4b0db8c2-9f26-4417-8bde-3f0e3656f8e0://auth` in the [authConfig.json](./redirect/public/config/authConfig.json) to forward the authorization code to the electron application. Deploying the static web page and using it as a redirect page for the electron app is recommended. For more information on hosting a static website on **Azure Storage**, please see [Host a static website on Blob Storage](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-blob-static-website-host)
```

```javascript
 "use static";

import auth from "../config/auth.json" assert { type: "json" };

window.addEventListener("DOMContentLoaded", function () {
    let redirectUri = `${auth.customURLScheme}`;

    if (window.location.search) {
        redirectUri += window.location.search;
        window.history.pushState({}, document.title, "/");
    }

    document.getElementById("SignIn").onclick = function () {
        window.location.href = redirectUri;
    };
});

```

### Deep linking

> To register the custom URL scheme and to link to the electron app instance. we will utilize the `setAsDefaultProtocolClient()` API provided by electron. For more information about using deep linking with electron, check [electron documentation](https://www.electronjs.org/docs/latest/tutorial/launch-app-from-url-in-another-ap)

Reviewer: ```suggestion
To register the custom URL scheme and to link to the electron app instance, we will utilize the `setAsDefaultProtocolClient()` API provided by electron. For more information about using deep linking with electron, check [electron documentation](https://www.electronjs.org/docs/latest/tutorial/launch-app-from-url-in-another-ap)
```

    ): void {
        // Someone tried to run a second instance, we should focus our window.
        if (Main.mainWindow) {
            if (Main.mainWindow.isMinimized()) Main.mainWindow.restore();
            Main.mainWindow.focus();
        }

        //get the deep linked from command line params
        const url = Main.getDeepLinkUrl(commandLine);

        if (url) {
            Main.getCodeFromDeepLinkUrl(url);
        }
    }
```

#### MacOS and Linux

When the external custom URL scheme is clicked it will trigger the event `Main.application.on("open-url", Main.onOpenUrl);` where we can fetch the authorization code from custom URL scheme:

```typescript
 private static onOpenUrl(event: any, schemeData: string) {
        event.preventDefault();
        // Someone tried to run a second instance, we should focus our window.
        if(Main.mainWindow){
            if (Main.mainWindow.isMinimized()) Main.mainWindow.restore();
            Main.mainWindow.focus();
        }

        //get the deep linked from command line params
        const url = Main.getDeepLinkUrl(schemeData);

         if (url) {
             Main.getCodeFromDeepLinkUrl(url);
         }

    }
```

> **Note: On macOS and Linux, this feature will only work when your app is packaged. It will not work when you're launching it in development from the command-line. For more information on how to pack an electron application please check the [electron documentation](https://www.electronjs.org/docs/latest/tutorial/launch-app-from-url-in-another-app#packaging>).**

Reviewer: @salman90 can you confirm this? I was able to run it directly on my macOS : )

Author: In my case, I had to pack the application, But I updated the note's description to "You May need to pack the app."  just in case it worked or didn't work in development.
